physical/etcd3.go

```diff
@@ -0,0 +1,285 @@
+		tls := transport.TLSInfo{
+			CAFile:   ca,
+			CertFile: cert,
+			KeyFile:  key,
+		}
+
+		tlscfg, err := tls.ClientConfig()
+		if err != nil {
+			return nil, err
+		}
+		cfg.TLS = tlscfg
+	}
+
+	// Set credentials.
+	username := os.Getenv("ETCD_USERNAME")
+	if username == "" {
+		username, _ = conf["username"]
+	}
+
+	password := os.Getenv("ETCD_PASSWORD")
+	if password == "" {
+		password, _ = conf["password"]
+	}
+
+	if username != "" && password != "" {
+		cfg.Username = username
+		cfg.Password = password
+	}
+
+	etcd, err := clientv3.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	sync, ok := conf["sync"]
+	if !ok {
+		sync = "yes"
+	}
+	switch sync {
+	case "yes", "true", "y", "1":
+		ctx, cancel := context.WithTimeout(context.Background(), client.DefaultRequestTimeout)
+		err := etcd.Sync(ctx)
+		cancel()
+		if err != nil {
+			return nil, err
+		}
+	case "no", "false", "n", "0":
+	default:
+		return nil, fmt.Errorf("value of 'sync' could not be understood")
+	}
+
+	return &Etcd3Backend{
+		path:       path,
+		etcd:       etcd,
+		permitPool: NewPermitPool(DefaultParallelOperations),
+		logger:     logger,
+		haEnabled:  haEnabledBool,
+	}, nil
+}
+
+func (c *Etcd3Backend) Put(entry *Entry) error {
+	defer metrics.MeasureSince([]string{"etcd", "put"}, time.Now())
+
+	c.permitPool.Acquire()
+	defer c.permitPool.Release()
+
+	_, err := c.etcd.Put(context.Background(), path.Join(c.path, entry.Key), string(entry.Value))
+	return err
+}
+
+func (c *Etcd3Backend) Get(key string) (*Entry, error) {
+	defer metrics.MeasureSince([]string{"etcd", "get"}, time.Now())
+
+	c.permitPool.Acquire()
+	defer c.permitPool.Release()
+
+	resp, err := c.etcd.Get(context.Background(), path.Join(c.path, key))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.Kvs) == 0 {
+		return nil, nil
+	}
+	return &Entry{
+		Key:   key,
+		Value: resp.Kvs[0].Value,
+	}, nil
+}
+
+func (c *Etcd3Backend) Delete(key string) error {
+	defer metrics.MeasureSince([]string{"etcd", "delete"}, time.Now())
+
+	c.permitPool.Acquire()
+	defer c.permitPool.Release()
+
+	_, err := c.etcd.Delete(context.Background(), path.Join(c.path, key))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Etcd3Backend) List(prefix string) ([]string, error) {
+	defer metrics.MeasureSince([]string{"etcd", "list"}, time.Now())
+
+	c.permitPool.Acquire()
+	defer c.permitPool.Release()
+
+	prefix = path.Join(c.path, prefix)
+	resp, err := c.etcd.Get(context.Background(), prefix, clientv3.WithPrefix())
+	if err != nil {
+		return nil, err
+	}
+
+	keys := []string{}
+	for _, kv := range resp.Kvs {
+		key := strings.TrimPrefix(string(kv.Key), prefix)
+		key = strings.TrimPrefix(key, "/")
+
+		if len(key) == 0 {
+			continue
+		}
+
+		if i := strings.Index(key, "/"); i == -1 {
+			keys = append(keys, key)
+		} else if i != -1 {
+			keys = appendIfMissing(keys, key[:i+1])
+		}
+	}
+	return keys, nil
+}
+
+func (e *Etcd3Backend) HAEnabled() bool { return e.haEnabled }
```

Comment:
Minor style nitpick: we don't do in-line functions...can you put this on the next line?